### ISSUE TITLE
Graph repository needs to tolerate fully escaped find strings

### ISSUE THREAD

**Issue by grahamwallis** [odpi/egeria#1597]

**Comment by cmgrote** [comment]
As discussed, have a look at the following helper methods:

- `OMRSRepositoryContentHelper::isExactMatchRegex(String)`
- `OMRSRepositoryContentHelper::isContainsRegex(String)`
- `OMRSRepositoryContentHelper::isStartsWithRegex(String)`
- `OMRSRepositoryContentHelper::isEndsWithRegex(String)`

These will tell you which "simple" style regular expression is being used (still a perfectly valid Java regex, but will have embedded `\Q` and `\E` quote characters that are Java-specific and may not be used in all regex interpretations).  Once you know what you're handling, you can then use this method to strip off the Java-specific pieces and get just the literal string:

- `OMRSRepositoryContentHelper::getUnqualifiedLiteralString(String)`

Just keep in mind that this strips off _all_ of the surrounding regex context.  So, for example:

- if `isContainsRegex(String)` returns true, it means the string looks something like `.*\Qsomething.you.should.treat.as.a.literal\E.*`
- which after sending to `getUnqualifiedLiteralString(String)` will give back just `something.you.should.treat.as.a.literal`
- so you need to remember whether you're treating it as a contains, startswith, or endswith before un-qualifying it
- (and also keep in mind that in this case all the `.`s in `something.you.should.treat.as.a.literal` should be treated as the literal character `.` and not as a regex placeholder for a single character)

**Comment by grahamwallis** [comment]
Hi @cmgrote - Raluca sent me a debug trace and it is as we suspected - the OMAS is calling getExactMatchRegex() which is 'framing' the search string with \Q \E. This is what is passed to the repo connector. I'm familiar with your constructor and destructor methods which are neat.

For the graph repo I have access to the regex predicates supported by Janus - and I'm trying tp work out how to get those to take heed of the \Q and \E framing. It seems they don't - which would mean that I need to strip them - which is easy but then would mean that any meta-characters in the stripped (unqualified) string would regain special significance. 
So for a non-special search string (e.g. any alphanumerics) life is good. But if we want to search literally for a qualifiedName containing 'ab.cd' then I have a problem with the '.' because if I simply unqualified the string then it would match any 5 character string starting 'ab' and ending 'cd'. I would need to re-escape just the special characters. 

Will keep investigating and trying to come up with a solution.

**Comment by cmgrote** [comment]
Yes, unfortunately this is the kind of thing every engine will need to do to translate Java-specific regex into its own regex language.  (In particular, to recognise whether it can support all of the Java-specific regex characters and meanings in its own regex language...)

If there are other helpers that make sense to add more generally to achieve this that could be handy; I'm just wary that there are many different forms of regex which are all subtly different, so not too keen to try to come up with some generic translation mechanism between all of them...

(Hence going back to my previous comments about the most common scenarios / use cases to know which helpers are likely to have the highest priority...)

**Comment by grahamwallis** [comment]
I think one key problem is our ability to capture the intent of the OMAS so that the repository connector understands what was intended and can do the right thing.
Suppose as an example we have an entity with qualfiiedName 'abide', another entity with qualifiedName 'abode' and a further entity with qualifiedName 'ab.de'.

The OMAS user/writer may want 'find' to retrieve only the 'abide' entity or 'abode' entity. This is an easy case - they could use the exact match helper to generate \Qabide\E or \Qabode\E and the repo connector could strip it using the getUnqualified..helper method'. The search would only retrieve the desired entity.

The harder case is where the OMAS user/writer wants to 
EITHER:
1. retrieve all 3 entities. This is what they would get if they used 'ab.de' as the search string, whether we frame it in \Q \E or not, since the graph connector seems to need to remove the framing.
OR
2. retrieve just the specific entity with name 'ab.de'. In this case we would need the OMAS to explicitly escape the '.' character - i.e. using 'ab\.de' then the repo connector could do the right search.

If they intend the first behaviour life is good; but if they want to specifically retrieve the entity whose name contains the special character we will currently retrieve too many entities, unless the graph connector can ascertain that was not the intention.

It seems logical in case 1 that the OMAS would not call a helper method  - since it is consciously providing a regex and the '.' character has special significance. And in the second case it seems logical that the OMAS would call an 'exact match' helper - which would individually literalise any special characters. 

Is that worth a go? (I'd be happy to write the literalising helper).

**Comment by mandy-chessell** [comment]
@grahamwallis I agree - that is what I coded in the AssetHelper.  I use the exact match helper method when retrieving a specific entity/relationship by qualified name.  If I want all of the assets matching the string then I just pass the search string.

see

**Comment by cmgrote** [comment]
The `getExactMatchRegex()` helper method _is_ literalising the special characters -- the way to do this in Java is to use [`java.util.regex.Pattern::quote(String)`](https://docs.oracle.com/javase/7/docs/api/java/util/regex/Pattern.html#quote(java.lang.String)), which is exactly how `getExactMatchRegex()` has been implemented.  (This is implemented natively in Java itself by adding the `\Q...\E` wrapping.)

Individually literalising / escaping every single potentially-significant character is going be error-prone and resource-intensive.  It also won't solve the issue where how something is specified in a Java regex differs from how the same regex would be defined in PCRE, JavaScript, Python, Golang or any other variations that a given repository back-end may or may not require.  Furthermore, for repositories that need to translate the regex into some other kind of predicate (contains, startswith, endswith, etc), we'd then need to do the reverse (also resource-intensive) to strip all of these escapes back off again.

So in a nutshell I'd be against this: either
- a repository supports all of Java regex (and can take the string as-is and doesn't need the helpers), like the in-memory repository
- or it doesn't support regex at all, but provides some predicates that can be easily translated to / from simple regexes
- or it only supports some other subset of Java regexes -- in which case it's up to that repository to parse the string and see whether it can or cannot support it

**Comment by grahamwallis** [comment]
OK - I think we are all saying the same things - I'm starting to write a specific literalizer. The existing helpers are good provided one stays on the Java island - but at the edge I need to do the translation into what JanusGraph supports. We're just going to have to cop the extra compute cost.
My point above about capturing the 'intent' of the OMAS may be sufficiently served by looking at whether the search string is framed by some combination of \Q \E and/or prefix/suffixed it with wildcards. My code will need to look at the string (for which I can use the isXXX helper methods) and will have to assert that if it finds the 'exact match' (QE) framing then the OMAS intended an exact match. There's no guarantee that's true but I think it's a reasonable guess. If an OMAS (or anyone) has explicitly framed their string with QE escapes then it's probably also fair to assume that they intended to. My connector will then construct the search string it needs.

**Comment by cmgrote** [comment]
> OK - I think we are all saying the same things - I'm starting to write a specific literalizer.

I think you're right -- apologies if I took your original suggestion the wrong way (to replace the existing `getExactMatchRegex()` helper), when what you really meant is defining a _separate_ helper which will escape all the special characters!  Indeed some repositories (like the graph repo) can use this to appropriately construct a search string that's specific to such repositories that support some subset of regexes (by which I mean, since it doesn't support the `\Q...\E` construct it by definition only supports a subset of Java regex).

> My code will need to look at the string (for which I can use the isXXX helper methods) and will have to assert that if it finds the 'exact match' (QE) framing then the OMAS intended an exact match. There's no guarantee that's true but I think it's a reasonable guess.

I'm not sure why this isn't guaranteed?  If a string has been wrapped with `\Q...\E` it is very specifically saying treat everything within it as an exact match.  The only risk would be someone calling the metadata collection method without sending it a regex -- but that violates the methods' documented interfaces.  Furthermore, as I understand it, we more or less only ever expect an OMAS to call the metadata collection methods (no end-user / application calling directly to OMRS), so I think it's up to us to enforce that OMASes are following the OMRS interface appropriate and only ever sending regex strings (ie. properly escaping them or not depending no the semantic they want to achieve?)

**Comment by cmgrote** [comment]
Ah, thinking about this a bit more maybe there are some cases where it isn't so simple, like:

`\Qsome.thing\Eand.some.more\E` -- not sure how this _should_ be interpreted (?)

But this also brings up the point that your literaliser is probably going to need to handle these `\Q...\E` pieces in a general way -- that is, the following is a perfectly legitimate Java regex:

`.*\Qtake.exact\Eand.not.exact\Qthen.more.exact\E.*`

**Comment by cmgrote** [comment]
Just a couple more checks for some concerns on my side:

1. We've found one example of Java regex characters that we can't support in graph repo (`\Q...\E`), but I think we need a definitive list so that the graph repo can check the string it receives to determine whether it can support it or needs to throw an exception?
1. Related to (1), presumably for some of these (like the `\Q...\E`) we will translate them rather than throwing an exception: what's the best way of documenting which ones we'll "support" (natively or translated) and which we won't?
1. How does the graph repo actually apply the regex to its search?  Is it something native in JanusGraph, that comes with no performance penalty?  Or (purely for extreme illustration) would there be an ability to search directly against an index without any regex interpretation in ~1ms vs sending a processor into the graph to calculate regex match against every object and return only the matches that might take ~1s?  IMHO we need to ensure that the exact match case is as optimal as possible, as a number of OMASes (like Data Engine) will be making 100's of search requests for exact matches for even relatively simple ETL jobs (of which there could be 10 000's) -- even 1-10ms savings on each of these calls will add up to very significant overall savings...  So my concern is if we're treating everything as a regex all the time for code simplicity, but with a potential performance penalty, this could have very significant implications to overall solution performance...

**Comment by grahamwallis** [comment]
@cmgrote @popa-raluca @lcpopa I have coded up a change to the graph repository connector and unit tested it; will functionally test and post a PR on Monday.

**Comment by grahamwallis** [comment]
Please see PR #1615

**Comment by popa-raluca** [comment]
> Please see PR #1615

Data Engine OMAS works with these latest changes
